modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/sub.js

```diff
@@ -0,0 +1,25 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+export default function sub(string, ...args) {
+	if (typeof string !== 'string') {
+		throw new TypeError('the first argument must be a string');
+	}
+
+	const REGEX_SUB = /\{\s*([^|}]+?)\s*(?:\|([^}]*))?\s*\}/g;
+
+	return string.replace(REGEX_SUB, (match, key) =>
+		args[key] === undefined ? match : args[key]
+	);
+}
```

Comment:
I think we should keep the original function of sub, by keeping the second arg as an array. The reason for this is because by changing this we are essentially causing a breaking change. Liferay.Lang.sub is a public API and so i am sure customers are using it and places outside of liferay-portal.

Reply:
Hmm, makes sense, that whole setup was confusing/weird to me and might have even been bugged. Doesn't this only return one argument: `data = Array.prototype.slice.call(arguments, 1);`? 
```js
const arguments = ['foo', 'bar', 'baz']

sub('foo-x', arguments); // returns 'foo-foo'
```
This is my problem with switching contexts, but IIRC, that's why I removed that whole array handling. But I can keep it in, for safety.

Reply:
I believe when language keys get to the client, they transform from 'foo-x' to 'Foo {0}'.

Check out the original code working here, https://codesandbox.io/s/beautiful-elion-nli1ng?file=/src/index.js

Reply:
Then I remembered wrong, a lot of these deprecations going on at the same time 🤷 

Thanks for catching it <3

Reply:
And just FYI, most of the issues in this epic are aimed at removing AUI and YUI from the util, rather than deprecating the utility all together. I just noticed that in your other PR.  Let me know if you have any questions about it

Reply:
So far, some of them had no AUI/YUI in them, like this `sub` one. Shouldn't we still move them away from `aui-js-web`?

Reply:
Also, responding here to your [comment](https://github.com/liferay-frontend/liferay-portal/pull/1992#discussion_r832085259), to keep it in this one thread:
> but I don't think we actually want to deprecate this utility as it is a public API for developers to consume. Unless you saw somewhere specifically about deprecating this.

Previously, when we did these deprecations/modernizations the goal was to deprecate them so that nothing breaks.

I think that our ultimate goal is to reduce the amount of utils, but we can't straight up remove them, so we deprecate them, then move the entire module to `archived` and call it a day (after a few years ofc). 

If a customer is using a util, nothing happens to them, everything stays the same as far as they're concerned, but we have moved/deprecated/modernized them, so that our life is easier.

Reply:
@bryceosterhaus pinging for above

Reply:
> @kresimir-coko
> if a utility isn't used anywhttps://github.com/liferay-frontend/liferay-portal/pull/1992 or in one to two places, deprecate it and rewrite its usage close to source (doing this here)

So we can't actually rely just on where it is used in `liferay-portal` because they are public APIs for customers to use. So we should expect that these utils are used frequently in customer code as well, which means we shouldn't just deprecate based on how many times it is used in liferay-portal. This also goes back to the original purpose of the epic, which is not about deprecating utilities, it is focused on removing AUI.

> @kresimir-coko 
> I think that our ultimate goal is to reduce the amount of utils

I don't think this is actually the goal of the epic. The goal of the epic is to remove AUI.
From the epic, the goal is stated:

> LPS-113522
> to remove all dependency on AUI so that we can turn off frontend-js-aui-web by default, and for any functionality that needs to live on, we want it ported into a "modernized" format (ie. remove dependency on legacy APIs and libraries, pay off technical debt etc).

And from the this task in particular, the goal is stated

> LPS-113543
> Purpose of this task is to contribute to our goal of removing our dependency on AUI/YUI from liferay portal

Deprecating a util means we intend to entirely remove it in the future, which for most of these utils, we don't actually plan on removing them at all, rather we just want to remove the dependency on AUI.


So in conclusion, these tasks should be focused on AUI removal and moving them out of aui-web, if necessary. Some of these utils may not even need changes, but we at least need to verify if they do.

Reply:
Well put @bryceosterhaus, makes sense, I'll switch up my approach 👍

Reply:
Sounds good! I am happy to talk through it more too if needed, we can always jump on a call as well.

Reply:
I'll see with the next couple tasks and if I have more questions then we can, for sure 😄